src/freertos_drivers/common/WiFiInterface.hxx

```diff
@@ -0,0 +1,377 @@
+    /// @param timeout is the number of milliseconds to wait for the WPS
+    /// process to complete. A value of zero indicates that the interface
+    /// default timeout should be used.
+    ///
+    /// @return @ref WiFiErrorCode as the result of the connection attempt.
+    virtual WiFiErrorCode initiate_wps_connect(bool blocking = false,
+                                               const uint32_t timeout = 0) = 0;
+
+    /// @return the current WiFi interface operating mode.
+    virtual WiFiMode get_mode() = 0;
+
+    /// @return the current WiFi interface status.
+    virtual WiFiState get_state() = 0;
+
+    /// Adds a saved SSID profile.
+    ///
+    /// @param ssid SSID of the profile to save.
+    /// @param password password of the SSID (preshared key), a blank string
+    /// is only valid for @ref WiFiSecurity::OPEN.
+    /// @param security @ref WiFiSecurity of the profile to be saved.
+    /// @param priority connection priority when more than one of the saved
+    /// networks is available, 0 == lowest priority
+    /// @param profile_index will contain the index of the newly stored
+    /// profile, if successfully stored.
+    ///
+    /// @return @ref WiFiErrorCode as the result of storing the profile.
+    virtual WiFiErrorCode add_ssid_profile(const std::string &ssid,
+                                           const std::string &password,
+                                           const WiFiSecurity security,
+                                           const uint8_t priority,
+                                           uint8_t *profile_index) = 0;
+
+    /// Delete a saved SSID profile by it's SSID.
+    ///
+    /// @param ssid SSID of the profile to delete.
+    ///
+    /// @return @ref WiFiErrorCode as the result of the removal of the profile.
+    virtual WiFiErrorCode del_ssid_profile(const std::string &ssid) = 0;
+
+    /// Delete a saved SSID profile by it's index.
+    ///
+    /// @param index index within saved profile list to remove, a value of 0xFF
+    /// will remove all stored profiles.
+    ///
+    /// @return @ref WiFiErrorCode as the result of the removal of the profile.
+    virtual WiFiErrorCode del_ssid_profile(const int index) = 0;
+
+    /// Retrieves a saved SSID profile by index.
+    ///
+    /// @param index of the stored profile to retrieve.
+    /// @param ssid will contain the ssid of the profile.
+    /// @param security will contain the security type of the profile.
+    /// @param priority will contain the priority of the profile.
+    ///
+    /// @return @ref WiFiErrorCode as the result of the retrieval of the
+    /// profile.
+    virtual WiFiErrorCode get_ssid_profile(const int index, std::string *ssid,
+                                           WiFiSecurity *security,
+                                           uint8_t *priority) = 0;
+
+    /// @return true if there are no saved SSID profiles, false otherwise.
+    virtual bool are_ssid_profiles_empty() = 0;
```

Comment:
Maybe instead of a boolean API, how about a num_ssid_profiles().  It can then be used in a loop counter.  This would require some abstraction in the CC32xx because indexes can be "skipped", but it would be much more user friendly at the application layer.

Reply:
Yes, switching to ssid_profile_count